File: src/lib/data/name/NameElement.cpp
#include "NameElement.h"

#include "logging.h"
#include "utilityString.h"

NameElement::Signature::Signature()
	: m_prefix(L"")
	, m_postfix(L"")
{
}

NameElement::Signature::Signature(std::wstring prefix, std::wstring postfix)
	: m_prefix(std::move(prefix))
	, m_postfix(std::move(postfix))
{
}

std::wstring NameElement::Signature::qualifyName(const std::wstring& name) const
{
	if (!isValid())
	{
		return name;
	}

	std::wstring qualifiedName = m_prefix;
	if (!name.empty())
	{
		if (!m_prefix.empty())
		{
			qualifiedName += L" ";
		}
		qualifiedName += name;
	}
	qualifiedName += m_postfix;

	return qualifiedName;
}

bool NameElement::Signature::isValid() const
{
	return ((m_prefix + m_postfix).size() > 0);
}

const std::wstring& NameElement::Signature::getPrefix() const
{
	return m_prefix;
}

const std::wstring& NameElement::Signature::getPostfix() const
{
	return m_postfix;
}

std::wstring NameElement::Signature::getParameterString() const
{
	if (m_postfix.size())
	{
		return utility::substrBeforeLast(m_postfix, L')') + L')';
	}

	return m_postfix;
}

NameElement::NameElement(std::wstring name)
	: m_name(std::move(name))
{
}

NameElement::NameElement(std::wstring name, std::wstring prefix, std::wstring postfix)
	: m_name(std::move(name))
	, m_signature(std::move(prefix), std::move(postfix))
{
}

NameElement::~NameElement()
{
}

const std::wstring& NameElement::getName() const
{
	return m_name;
}

std::wstring NameElement::getNameWithSignature() const
{
	return m_signature.qualifyName(m_name);
}

std::wstring NameElement::getNameWithSignatureParameters() const
{
	return m_name + m_signature.getParameterString();
}

bool NameElement::hasSignature() const
{
	return m_signature.isValid();
}

const NameElement::Signature& NameElement::getSignature()
{
	return m_signature;
}

File: java_indexer/src/main/java/com/sourcetrail/name/NameElement.java
package com.sourcetrail.name;

public class NameElement
{
	private String m_name = "";
	private String m_prefix = "";
	private String m_postfix = "";


	public NameElement(String name)
	{
		if (name != null)
			m_name = name;
	}

	public NameElement(String name, String prefix, String postfix)
	{
		if (name != null)
			m_name = name;
		if (prefix != null)
			m_prefix = prefix;
		if (postfix != null)
			m_postfix = postfix;
	}

	String getName()
	{
		return m_name;
	}

	String getNameWithSignature()
	{
		String nameWithSignature = m_name;

		if (m_prefix.length() > 0 || m_postfix.length() > 0)
		{
			nameWithSignature = m_prefix;
			if (!m_name.isEmpty())
			{
				if (!m_prefix.isEmpty())
				{
					nameWithSignature += " ";
				}
				nameWithSignature += m_name;
			}
			nameWithSignature += m_postfix;
		}

		return nameWithSignature;
	}

	String serialize()
	{
		return m_name + "\ts" + m_prefix + "\tp" + m_postfix;
	}
}
